Add support for SKY Wallet

skywalletadmin opened the pull request:
This PR adds support for SKY Wallet
Currently SKY Wallet works as a DApp Browser Implementation inside the SKY Wallet mobile app. Support for Chrome and Safari coming soon.

jordansexton commented:
Hey, I want to address the lack of activity on your PR. Thank you for your patience!

Here's a followup on the state of wallet-adapter, the issues that individual adapters have in getting integrated with applications, and how we intend to improve things: https://github.com/solana-labs/wallet-adapter/pull/384#issuecomment-1159115764

Let's discuss in that thread to keep it in one place.

jordansexton commented on packages/starter/example/components/ContextProvider.tsx:
```diff
@@ -1,22 +1,24 @@
-import { WalletModalProvider as AntDesignWalletModalProvider } from '@solana/wallet-adapter-ant-design';
-import { WalletAdapterNetwork, WalletError } from '@solana/wallet-adapter-base';
-import { WalletDialogProvider as MaterialUIWalletDialogProvider } from '@solana/wallet-adapter-material-ui';
+import { AutoConnectProvider, useAutoConnect } from './AutoConnectProvider';
```
Please limit changes to adding the package only. We don't add all new wallets to the example or starter projects.

jordansexton replied:
It doesn't look like this was fixed.

jordansexton commented on packages/wallets/skywallet/package.json:
```diff
@@ -0,0 +1,39 @@
+{
+    "name": "@solana/wallet-adapter-skywallet",
```
Can this be `@solana/wallet-adapter-sky`?

jordansexton commented on packages/wallets/skywallet/package.json:
```diff
@@ -0,0 +1,39 @@
+        "@solana/wallet-adapter-base": "^0.9.4"
+    },
+    "devDependencies": {
+        "@solana/wallet-adapter-base": "^0.9.5",
```
This shouldn't be declared as a dev dep.

jordansexton replied:
Can you make it so this matches the peer/dev/deps shown here? https://github.com/solana-labs/wallet-adapter/blob/244af02cb3dcf9334abf61cb9a11e497bde86d57/packages/wallets/phantom/package.json#L29-L37

jordansexton commented on packages/wallets/skywallet/src/adapter.ts:
```diff
@@ -0,0 +1,262 @@
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const wallet = window!.skySolana!;
+
+            if (!wallet.isConnected) {
+                // HACK: SkyWallet doesn't reject or emit an event if the popup is closed
+                const handleDisconnect = wallet._handleDisconnect;
+                try {
+                    await new Promise<void>((resolve, reject) => {
+                        const connect = () => {
+                            wallet.off('connect', connect);
+                            resolve();
+                        };
+
+                        wallet._handleDisconnect = (...args: unknown[]) => {
+                            wallet.off('connect', connect);
+                            reject(new WalletWindowClosedError());
+                            return handleDisconnect.apply(wallet, args);
+                        };
+
+                        wallet.on('connect', connect);
+
+                        wallet.connect().catch((reason: any) => {
+                            wallet.off('connect', connect);
+                            reject(reason);
+                        });
+                    });
+                } catch (error: any) {
+                    if (error instanceof WalletError) throw error;
+                    throw new WalletConnectionError(error?.message, error);
+                } finally {
+                    wallet._handleDisconnect = handleDisconnect;
+                }
+            }
```
It looks like this was all copied from the Phantom adapter, which has hacks to access private APIs because of bad connection behavior. Please don't copy these things!

jordansexton commented on packages/wallets/skywallet/src/adapter.ts:
```diff
@@ -0,0 +1,262 @@
+        options?: SendTransactionOptions
+    ): Promise<TransactionSignature> {
+        try {
+            const wallet = this._wallet;
+            if (wallet && 'signAndSendTransaction' in wallet && !options?.signers) {
+                const { signature } = await wallet.signAndSendTransaction(transaction, options);
+                return signature;
+            }
```
Same here, please don't copy the Phantom adapter's workarounds. Partially signed transactions should also be supported.

jordansexton reviewed (commented):
While the [medium-term goal](https://github.com/solana-labs/wallet-adapter/pull/384#issuecomment-1159115764) is to move away from requiring adapters for every wallet, I want to respect the time you took to create this adapter, and review it now.

Once a new standard is introduced, let's work together to migrate away from dependency on using an adapter.

skywalletadmin commented:
Thanks Jordan for the review on this PR. We have resolved all your comments and pushed the new changes. Looking forward to seeing SKY Wallet integrated with this PR. Please let me know if any other changes are required. Thanks!

jordansexton commented on packages/wallets/skywallet/src/adapter.ts:
```diff
@@ -0,0 +1,261 @@
+        this.emit('disconnect');
+    }
+
+    async sendTransaction(
+        transaction: Transaction,
+        connection: Connection,
+        options: SendTransactionOptions = {}
+    ): Promise<TransactionSignature> {
+        let emit = true;
+        try {
+            try {
+                transaction.feePayer = transaction.feePayer || this.publicKey || undefined;
+                transaction.recentBlockhash =
+                    transaction.recentBlockhash || (await connection.getRecentBlockhash('finalized')).blockhash;
+
+                const { signers, ...sendOptions } = options;
+
+                signers?.length && transaction.partialSign(...signers);
+
+                transaction = await this.signTransaction(transaction);
+
+                const rawTransaction = transaction.serialize();
+
+                return await connection.sendRawTransaction(rawTransaction, sendOptions);
+            } catch (error: any) {
+                if (error instanceof WalletError) {
+                    emit = false;
+                    throw error;
+                }
+                throw new WalletSendTransactionError(error?.message, error);
+            }
+        } catch (error: any) {
+            if (emit) {
+                this.emit('error', error);
+            }
+            throw error;
+        }
+    }
```
It looks like this just reimplements this entire function: https://github.com/solana-labs/wallet-adapter/blob/244af02cb3dcf9334abf61cb9a11e497bde86d57/packages/core/base/src/signer.ts#L13-L48

Why? The interface for `SkyWallet above declares a `signAndSendTransaction` method. This should be used if available.

Please don't just copy code for your adapter, make sure that it correctly adapts the interface of your wallet.

jordansexton commented on packages/wallets/skywallet/src/adapter.ts:
```diff
@@ -0,0 +1,252 @@
+            try {
+                await wallet.connect();
+            } catch (error: any) {
+                if (error instanceof WalletError) {
+                    throw error;
+                }
```
This check isn't needed because the error is thrown by the extension (it'll never be a `WalletError` instance).

jordansexton commented on packages/wallets/skywallet/src/adapter.ts:
```diff
@@ -0,0 +1,252 @@
+                const { signature } = await wallet.signMessage(message);
+                return signature;
+            } catch (error: any) {
+                throw new WalletSignTransactionError(error?.message, error);
```
I just caught this in several other adapters today, but this should be `WalletSignMessageError`

https://github.com/solana-labs/wallet-adapter/commit/112c54463a3d069e5b756dacf78e20a0711939ae fixes this elsewhere

jordansexton commented on packages/wallets/wallets/package.json:
```diff
@@ -47,6 +47,7 @@
         "@solana/wallet-adapter-solong": "^0.9.2",
         "@solana/wallet-adapter-tokenpocket": "^0.4.2",
         "@solana/wallet-adapter-torus": "^0.11.12",
+        "@solana/wallet-adapter-sky": "^0.1.0",
```
nit: the packages should be alphabetical to prevent churn later

skywalletadmin commented:
Hi Jordan, 
Thanks for your review on this PR. We have addressed and fixed all the review comments.
Please let us know if there are any other changes to this. Much appreciated!

jordansexton commented on packages/wallets/skywallet/src/adapter.ts:
```diff
@@ -0,0 +1,248 @@
+            transaction.feePayer = transaction.feePayer || this.publicKey || undefined
+            transaction.recentBlockhash = transaction.recentBlockhash || (await connection.getRecentBlockhash('finalized')).blockhash;
+
+            if (options.signers?.length) {
+                // handling partial signing 
+                const { signers, ...sendOptions } = options;
+                transaction.partialSign(...signers);
+                transaction = await this.signTransaction(transaction);
+                const rawTransaction = transaction.serialize(); 
+  
+                const signature = await connection.sendRawTransaction(rawTransaction, sendOptions); 
+                return signature
+            }
```
This doesn't need to be reimplemented, does it? The base class already has all this, you can call the super method if you don't support partial signers.

jordansexton commented on packages/wallets/walletconnect/lib/cjs/adapter.js:
```diff
@@ -0,0 +1,283 @@
+"use strict";
```
These files shouldn't be included, please rebase / delete them.

jordansexton commented on packages/wallets/wallets/src/index.ts:
```diff
@@ -13,9 +13,12 @@ export * from '@solana/wallet-adapter-mathwallet';
 export * from '@solana/wallet-adapter-solong';
 export * from '@solana/wallet-adapter-tokenpocket';
 export * from '@solana/wallet-adapter-torus';
+export * from '@solana/wallet-adapter-walletconnect';
```
Merge master to get rid of this stuff

skywalletadmin commented:
Hi Jordan,
Thanks for your review! We've fixed the issues and made the changes. Please let us know if there's any thing else. Thanks!

jordansexton commented:
Published:
 - @solana/wallet-adapter-sky@0.1.0
 - @solana/wallet-adapter-wallets@0.16.5
